src/Servers/Kestrel/perf/Kestrel.Performance/SchedulerBenchmark.cs

```diff
@@ -0,0 +1,199 @@
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Performance
+{
+    public class SchedulerBenchmark
+    {
+        private const int InnerLoopCount = 1024;
+        private const int OuterLoopCount = 64;
+        private const int OperationsPerInvoke = InnerLoopCount * OuterLoopCount;
+
+        private readonly int IOQueueCount = Math.Min(Environment.ProcessorCount, 16);
+
+        private PipeScheduler[] _lockFreeQueue;
+        private PipeScheduler[] _lockBasedQueue;
+        private PipeScheduler _threadPoolScheduler;
+
+        private static Action<object> _action = (o) => { };
+
+        private static Action<int> _lockFreeAction;
+        private static Action<int> _lockBasedAction;
+        private static Action<int> _threadPoolAction;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _lockFreeQueue = new IOQueueLockFree[IOQueueCount];
+            for (var i = 0; i < _lockFreeQueue.Length; i++)
+            {
+                _lockFreeQueue[i] = new IOQueueLockFree();
+            }
+
+            _lockFreeAction =
+                (n) =>
+                {
+                    PipeScheduler pipeScheduler = _lockFreeQueue[n % _lockFreeQueue.Length];
+                    for (var i = 0; i < InnerLoopCount; i++)
+                    {
+                        pipeScheduler.Schedule(_action, null);
+                    }
+                };
+
+            _lockBasedQueue = new IOQueueLockBased[IOQueueCount];
+            for (var i = 0; i < _lockBasedQueue.Length; i++)
+            {
+                _lockBasedQueue[i] = new IOQueueLockBased();
+            }
+
+            _lockBasedAction =
+                (n) =>
+                {
+                    PipeScheduler pipeScheduler = _lockBasedQueue[n % _lockBasedQueue.Length];
+                    for (var i = 0; i < InnerLoopCount; i++)
+                    {
+                        pipeScheduler.Schedule(_action, null);
+                    }
+                };
+
+            _threadPoolScheduler = PipeScheduler.ThreadPool;
+            _threadPoolAction =
+                (n) =>
+                {
+                    for (var i = 0; i < InnerLoopCount; i++)
+                    {
+                        _threadPoolScheduler.Schedule(_action, null);
+                    }
+                };
+        }
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke, Baseline = true)]
+        public void LockBasedIOQueue() => Schedule(_lockBasedAction);
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
+        public void LockFreeIOQueue() => Schedule(_lockFreeAction);
+
+        [Benchmark(OperationsPerInvoke = OperationsPerInvoke)]
+        public void ThreadPoolDirect() => Schedule(_threadPoolAction);
+
+        private void Schedule(Action<int> scheduleAction) => Parallel.For(0, OuterLoopCount, scheduleAction);
```

Comment:
This doesn't seem to measuring how long it takes to actually execute all the scheduled actions, just how long it takes to schedule/enqueue it.

Can this test wait until the action Interlocked.Increments to `OuterLoopCount * InnerLoopCount` or something to verify that not only is scheduling complete, but what was scheduled is also complete?

Reply:
The contention was on scheduling for the IOQueue (lock per queue item, many threads queuing) rather than on executing, so that what I was looking at for the benchmark (i.e. to prove this PR)

However; having accepted the change to IOQueue, agreed that the benchmark makes more sense to compare the 3 variants (ThreadPool, IOQueue and Inline) and wait for completion.

Just working out a way to do it without hitting contention on the finish condition... give me a few.

Reply:
Done